pinot-core/src/main/java/org/apache/pinot/core/plan/SelectionPlanNode.java

```diff
@@ -50,67 +52,113 @@ public Operator<IntermediateResultsBlock> run() {
-      } else {
-        // Selection order-by
-        if (isAllOrderByColumnsSorted(orderByExpressions)) {
-          // All order-by columns are sorted, no need to sort the records
-          TransformOperator transformOperator = new TransformPlanNode(_indexSegment, _queryContext, expressions,
-              Math.min(limit + _queryContext.getOffset(), DocIdSetPlanNode.MAX_DOC_PER_CALL)).run();
-          return new SelectionOrderByOperator(_indexSegment, _queryContext, expressions, transformOperator, true);
-        } else if (orderByExpressions.size() == expressions.size()) {
-          // All output expressions are ordered
-          TransformOperator transformOperator =
-              new TransformPlanNode(_indexSegment, _queryContext, expressions, DocIdSetPlanNode.MAX_DOC_PER_CALL).run();
-          return new SelectionOrderByOperator(_indexSegment, _queryContext, expressions, transformOperator, false);
-        } else {
-          // Not all output expressions are ordered, only fetch the order-by expressions and docId to avoid the
-          // unnecessary data fetch
-          List<ExpressionContext> expressionsToTransform = new ArrayList<>(orderByExpressions.size());
-          for (OrderByExpressionContext orderByExpression : orderByExpressions) {
-            expressionsToTransform.add(orderByExpression.getExpression());
-          }
-          TransformOperator transformOperator =
-              new TransformPlanNode(_indexSegment, _queryContext, expressionsToTransform,
-                  DocIdSetPlanNode.MAX_DOC_PER_CALL).run();
-          return new SelectionOrderByOperator(_indexSegment, _queryContext, expressions, transformOperator, false);
-        }
-      }
-    } else {
+    if (limit == 0) {
       // Empty selection (LIMIT 0)
       TransformOperator transformOperator = new TransformPlanNode(_indexSegment, _queryContext, expressions, 0).run();
       return new EmptySelectionOperator(_indexSegment, expressions, transformOperator);
     }
+    List<OrderByExpressionContext> orderByExpressions = _queryContext.getOrderByExpressions();
+
+    if (orderByExpressions == null) {
+      // Selection only
+      // ie: SELECT ... FROM Table WHERE ... LIMIT 10
+      int actualLimit = Math.min(limit, DocIdSetPlanNode.MAX_DOC_PER_CALL);
+      TransformPlanNode planNode = new TransformPlanNode(_indexSegment, _queryContext, expressions, actualLimit);
+      TransformOperator transformOperator = planNode.run();
+
+      return new SelectionOnlyOperator(_indexSegment, _queryContext, expressions, transformOperator);
+    }
+    int numOrderByExpressions = orderByExpressions.size();
+    // Although it is a break of abstraction, some code, specially merging, assumes that if there is an order by
+    // expression the operator will return a block whose selection result is a priority queue.
+    int sortedColumnsPrefixSize = getSortedColumnsPrefix(orderByExpressions);
```

Comment:
I don't get it. What do you mean with _null value can be stored as any default value_? The SQL standard doesn't define whether `null` is considered smaller or larger than non-null values (and offers the `ORDER BY <x> <ASC/DESC> NULL <FIRST/LAST>` to deal with it) but I guess we should be consistent.

If Pinot were added to this page, which mode would it use? https://learnsql.com/blog/how-to-order-rows-with-nulls/

Reply:
I believe we use `LAST` by default to match the Presto behavior: https://prestodb.io/docs/current/sql/select.html#order-by-clause

Currently we store `null` as any default value, so there can be a corner case where a sorted column has null values inside. Per the convention we pick, column with null value should not be count as sorted, so we should account for that

Reply:
Let me verify if I correctly understand it. Given a table `T` with three rows where with a single column called `col` whose values are `1`, `null` and `2`.

1. `col` is considered sorted.
2. If the default null value of `col` is null, `select col from T order by T` should return `1`, `2`, `null`

Is that right? And if it is... why do we considered `col` a sorted column if, in fact, it is not sorted in the segment? More formally, given for each d1 and d2 such as d1 < d2 then it is not guaranteed that segment[d1][col] <= segment[d2][col]

Reply:
It might be sorted or not based on the default value of the column. E.g. if default value for a int column is `0`, then `-1, null, 1` is considered sorted; but if the default value is `-100` then it is not sorted.

The null handling support feature is added recently, and we treat null as default value when generating the index (we don't know if null handling is enabled at ingestion time either). So what we can do here is deciding whether to use this optimization at query time, and not use it when null handling is enabled and the column contains null values

Reply:
> So what we can do here is deciding whether to use this optimization at query time, and not use it when null handling is enabled and the column contains null values

What I did was to change the method that checks whether an expression is sorted or not by adding the following condition:
```java
        // If there are null values, we cannot trust DataSourceMetadata.isSorted
        if (isNullHandlingEnabled) {
          NullValueVectorReader nullValueVector = dataSource.getNullValueVector();
          if (nullValueVector != null && !nullValueVector.getNullBitmap().isEmpty()) {
            return false;
          }
        }
```

AFAIK that is the most precise way to check if the column has a null value. I'm open to change that to other better ways if exists.

It would be great if we could do something like that directly in DataSourceMetadata, but we don't have the null vector there.